lib/index.js

```diff
@@ -55,82 +55,79 @@ function setupRateLimitEventExt(server, options) {
-          newMinimumLimitResponse.limit,
-          newMinimumLimitResponse.remaining,
-          newMinimumLimitResponse.reset);
-
-        reply(error);
-      });
-    });
+  const extractKeyAndTakeToken = async function(limitd, request, handler, type) {
+    const key = extractKey(request, handler);
+    if (err) { return err; }
+
+    if (!limitd) {
+      // limitd is not connected, do not fail!
+      return handler.continue;
+    }
+
+    try {
+      const currentLimitResponse = await limitd.take(type, key);
+
+    } catch(err) {
+      if (onError) { return onError(err, reply); }
+      // by default we don't fail if limitd is unavailable
+      return handler.continue;
+    }
+
+
+    const oldMinimumLimitResponse = request.plugins.patova && request.plugins.patova.limit
+    const newMinimumLimitResponse = getMinimumLimit(currentLimitResponse, oldMinimumLimitResponse)
+
+    request.plugins.patova = request.plugins.patova || {};
+    request.plugins.patova.limit = newMinimumLimitResponse;
+
+    if (newMinimumLimitResponse.conformant) {
+      // We continue only if the request is conformat so far
+      return handler.continue;
+    }
+
+    const error = Boom.tooManyRequests();
+    error.output.headers = new RateLimitHeaders(
+      newMinimumLimitResponse.limit,
+      newMinimumLimitResponse.remaining,
+      newMinimumLimitResponse.reset);
+
+    throw error;
   };
 
-  const getType = function(request, reply, callback) {
+  const getType = async function(request, handler, callback) {
     const type = options.type;
 
     if (typeof type !== 'function') {
-      return process.nextTick(() => callback(null, type));
+      return process.nextTick(() => Promise.resolve(type));
     }
 
     try {
-      return type(request, (err, type) => {
-        if (err) {
-          return reply(Boom.wrap(err, 500, 'cannot get bucket type'));
-        }
-
-        callback(null, type);
-      });
+      return type(request);
     } catch (err) {
       return reply(Boom.wrap(err, 500, 'cannot get bucket type'));
     }
   };
 
-  server.ext(event, (request, reply) => {
+  server.ext(event, async (request, handler) => {
     // This handler is going to be called one time per registration of patova
-    getType(request, reply, (err, type) => {
-      extractKeyAndTakeToken(options.limitd, request, reply, type);
-    });
+    const type = await getType(request, reply);
+    extractKeyAndTakeToken(options.limitd, request, handler, type);
   });
 }
 
-exports.register = function (server, options, next) {
-  Joi.validate(options, schema, { abortEarly: false }, (err, processedOptions) => {
-    if (err) { return next(err); }
+async function register(server, options) {
+  try {
+    const processedOptions = await Joi.validate(options, schema, { abortEarly: false });
     setupRateLimitEventExt(server, processedOptions);
     setupPreResponseExt(server, processedOptions);
-    next();
-  });
+    return
+  } catch(e) {
+    return e;
```

Comment:
Not sure, but shouldn't we just throw the exception instead? ie. removing the try/catch.